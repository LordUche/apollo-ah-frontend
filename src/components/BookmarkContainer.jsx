/* eslint-disable react/jsx-indent */
/* eslint-disable react/jsx-closing-tag-location */
/* eslint-disable react/no-array-index-key */
/* eslint-disable react/jsx-wrap-multilines */
import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import BookmarkItem from '../views/BookmarkItem';
import { getBookmarkedArticles } from '../actions/getBookmark';
import * as selectors from '../selectors/getBookmarkSelector';

/**
 * @description - returns all articles bookmarked by a user
 * @return {JSX}
 */
function BookmarkContainer(props) {
  const [viewType, setViewType] = useState('row view-row');

  const { articles } = props;

  const getArticlesHandler = () => {
    props.getBookmarkedArticles();
  };

  const changeViewToHorizontal = () => {
    setViewType('row view-row list-view');
  };

  const changeViewToVertical = () => {
    setViewType('row view-row');
  };

  useEffect(() => (
    getArticlesHandler()
  ), []);

  return (
    <>
      <header className="page-header" data-pg-collapsed>
        <h1>Bookmarks </h1>
        <div className="btn-group btn-group-toggle float-right view-changer" data-toggle="buttons">
          <label className="btn active" onClick={changeViewToVertical}>
            <input type="radio" name="options" autoComplete="off" checked value="" />
            <i className="fas fa-th" />
          </label>
          <label className="btn" onClick={changeViewToHorizontal}>
            <input type="radio" name="options" autoComplete="off" value="list-view" />
            <i className="fas fa-list" />
          </label>
        </div>
      </header>
      {
        articles.articles.length !== 0
          ? (
            <>
              <div className={viewType}>
                {
                  // eslint-disable-next-line arrow-body-style
                  articles.articles.map((article, indx) => {
                    return (
                      <BookmarkItem
                        key={indx}
                        category={article.Article.articleCategory.category}
                        image={article.Article.image}
                        articleTitle={article.Article.title}
                        articleDescription={article.Article.description}
                        authorsName={article.Article.User.Profile.firstname}
                        authorsImage={article.Article.User.Profile.image}
                        readTime={article.Article.readTime}
                      />
                    );
                  })
                }
              </div>
              {/** pagination support to be implemented later */}
              <div>Pagination is here</div>
            </>
          ) : <div>
            <i className="fas fa-book-open" style={{ fontSize: '70px', color: '#66008c' }} />
            <p style={{ paddingTop: '30px' }}>*It looks like you have not bookmarked any article yet</p>
          </div>
      }
    </>
  );
}

const mapStateToProps = createStructuredSelector(
  {
    isLoading: selectors.getBookmarkLoading,
    articles: selectors.getBookmarkArticles,
  }
);

export default connect(mapStateToProps, { getBookmarkedArticles })(BookmarkContainer);
