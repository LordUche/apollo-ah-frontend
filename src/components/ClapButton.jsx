/* eslint-disable no-undef */
import React, { Component } from 'react';
import PropType from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { debounce } from 'lodash';
import { withRouter } from 'react-router-dom';
import clapIcon from '../images/clap.svg';
import * as articleSelector from '../selectors/singleArticleSelector';
import { getAuthId, getIsAuthenticated } from '../selectors/authSelector';
import { getToken } from '../selectors/profileSelector';
import { clapArticleRequest, fetchUserClaps } from '../actions/clapsAction';
import { getUserClaps } from '../selectors/clapsSelector';
import { CLAPS_LIMIT, CLAPS_REQUEST_TIMER } from '../utils/contants';

/**
 * A class representing clap button component
 *
 * @class ClapButton
 * @extends {Component}
 */
export class ClapButton extends Component {
  /**
   * Handles clap request (debounced)
   *
   * @returns {void}
   * @memberof ClapButton
   */
  makeClapRequest = debounce(() => {
    const { clapArticle, articleSlug: slug, token } = this.props;
    const { clapsCount: claps } = this.state;
    this.setState({ clapsCount: 0 });
    return clapArticle({ slug, claps, token });
  }, CLAPS_REQUEST_TIMER);

  static propTypes = {
    articleSlug: PropType.string.isRequired,
    articleAuthorId: PropType.number.isRequired,
    articleClaps: PropType.number.isRequired,
    clapArticle: PropType.func.isRequired,
    isLoggedIn: PropType.bool.isRequired,
    userId: PropType.number,
    history: PropType.object.isRequired,
    token: PropType.string.isRequired,
    userClaps: PropType.number.isRequired,
    loadUserClaps: PropType.func.isRequired,
  };

  static defaultProps = {
    userId: null,
  };

  constructor(props) {
    super(props);
    this.state = {
      clapsLeft: CLAPS_LIMIT - props.userClaps,
      articleClaps: props.articleClaps,
      clapsCount: 0, // record clap button clicks
    };
  }

  componentDidMount = async () => {
    const {
      articleSlug: slug,
      isLoggedIn,
      userId,
      loadUserClaps
    } = this.props;
    if (isLoggedIn) {
      loadUserClaps({ slug, userId });
    }
  }

  /**
   * Makes an article clap
   *
   * @returns {void}
   * @memberof ClapButton
   */
  handleClapArticle = () => {
    const {
      isLoggedIn,
      userId,
      articleAuthorId,
      history,
    } = this.props;

    if (!isLoggedIn) { // redirect guest users to login
      history.push('/login');
      return false;
    }
    // Deny article's author
    if (articleAuthorId === userId) return false;
    const { articleClaps, clapsLeft, clapsCount } = this.state;
    if (clapsLeft >= 1) {
      this.setState({
        clapsLeft: clapsLeft - 1,
        articleClaps: articleClaps + 1,
        clapsCount: clapsCount + 1,
      });
      this.makeClapRequest();
    }
  }

  /**
   * React render function
   *
   * @returns {JSX.Element} DOM element
   * @memberof ClapButton
   */
  render() {
    const { articleClaps } = this.state;

    return (
      <span className="clap-icon">
        <button type="button" className="btn" onClick={this.handleClapArticle}>
          <img src={clapIcon} alt="" />
        </button>
        {articleClaps}
        {' '}
        claps
      </span>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  articleSlug: articleSelector.getArticleSlug,
  articleAuthorId: articleSelector.getArticleAuthorId,
  articleClaps: articleSelector.getArticleClaps,
  isLoggedIn: getIsAuthenticated,
  userId: getAuthId,
  token: getToken,
  userClaps: getUserClaps,
});

const mapDispatchToProps = {
  clapArticle: claps => clapArticleRequest(claps),
  loadUserClaps: payload => fetchUserClaps(payload),
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(ClapButton));
