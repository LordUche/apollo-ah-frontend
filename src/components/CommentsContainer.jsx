/* eslint-disable react/prefer-stateless-function */
import React, { Component } from 'react';

import Comments from '../views/Comments';

/**
 * Container component for the Article comments view
 * @export
 * @class CommentsContainer
 * @extends {Component}
 */
export default class CommentsContainer extends Component {
  render() {
    return (
      <Comments />
    );
  }
}
