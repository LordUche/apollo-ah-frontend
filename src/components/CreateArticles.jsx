/* eslint-disable no-unused-vars */
/* eslint-disable arrow-parens */
/* eslint-disable react/jsx-no-bind */
/* eslint-disable jsx-a11y/no-redundant-roles */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable jsx-a11y/label-has-for */
import React, { useState } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import * as selectors from '../selectors/createArticleSelector';
import createArticleAction from '../actions/createArticles';
import CreateArticle from '../views/CreateArticle';
import CloudinaryWidget from './common/CloudinaryWidget';


/**
 * Form for creating a new article
 * @param {object} props Component props
 * @param {function} props.createArticleAction action creator function
 */
function CreateArticleContainer(props) {
  const [formData, setFormData] = useState({
    title: null,
    description: null,
    body: null,
    categoryId: null,
    tagList: [],
    image: null,
  });

  /** passing global state loading as props into variables */
  let { message, isLoading, token } = props;

  let widget;

  const updateInputHandler = (event) => {
    setFormData({ ...formData, [event.target.name]: event.target.value });
  };

  const updateBodyHandler = (value) => {
    setFormData({ ...formData, body: value });
  };

  const updateTagHandler = (tags) => {
    const listOfTags = [];
    tags.forEach(tag => {
      listOfTags.push(tag.text);
      setFormData({ ...formData, tagList: listOfTags });
    });
  };

  const submitHandler = (event) => {
    event.preventDefault();
     console.log('--->', formData);
     const requestData = {
       formData,
       token,
     };
    //props.createArticleAction(formData);
    props.createArticleAction(requestData);
  };

  const cloudinaryUpdate = (imageUrl) => {
    setFormData({ ...formData, image: imageUrl });
  };

  const imageUploadHandler = () => {
    // console.log('yo mama');
    // const { showLoader } = this.props;
    // showLoader(true);
    if (widget === undefined) {
      widget = new CloudinaryWidget(
        cloudinaryUpdate, (error) => {
        return (error);
      }, false);
    }
    widget.open();
  };

  console.log('--->', isLoading);
  console.log('==->', message);

  return (
    <CreateArticle
      updateInputHandler={updateInputHandler}
      updateBodyHandler={updateBodyHandler}
      updateTagHandler={updateTagHandler}
      imageUploadHandler={imageUploadHandler}
      submitHandler={submitHandler}
      valueText={formData.body}
    />
  );
}

const mapStateToProps = createStructuredSelector(
  {
    isLoading: selectors.getCreateArticleLoadingState,
    message: selectors.getCreateArticleMessageState,
    token: selectors.getUserToken,
  }
);

export default connect(mapStateToProps, { createArticleAction })(CreateArticleContainer);
