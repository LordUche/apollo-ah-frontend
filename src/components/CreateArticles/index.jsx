/* eslint-disable jsx-a11y/no-redundant-roles */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable jsx-a11y/label-has-for */
import React from 'react';

function CreateArticle(props) {
  return (
    <section className="wrapper min-vh-100">
      <div className="container">
        <header className="page-header" data-pg-collapsed>
          <h1> New Article</h1>
        </header>
        <div className="card mb-5 mt-n5">
          <form role="form" className="form p-sm-4 p-3 form-borderless">
            <div className="form-group mb-4" data-pg-collapsed>
              <label htmlFor="title" className="font-weight-bold">Title</label>
              <input type="text" className="form-control text-capitalize" id="title" name="title" required />
            </div>
            <div className="form-group mb-4" data-pg-collapsed>
              <label htmlFor="description" className="font-weight-bold">Short Description</label>
              <textarea className="form-control textarea-sm" id="description" name="description" />
            </div>
            <div className="form-group mb-4" data-pg-collapsed>
              <label htmlFor="description" className="font-weight-bold">Details</label>
              <textarea className="form-control" id="body" name="body" />
            </div>
            <div className="form-group mb-4" data-pg-collapsed>
              <label className="font-weight-bold" htmlFor="category">Category</label>
              <select className="custom-select" name="category" id="category">
                <option selected value="-1">-- Choose one --</option>
                <option value="3">Business</option>
                <option value="7">Entertainment</option>
                <option value="4">Fashion</option>
                <option value="6">Food</option>
                <option value="5">Health</option>
                <option value="8">History</option>
                <option value="1">Technology</option>
                <option value="2">Sports</option>
                <option value="9">Others</option>
              </select>
            </div>
            <div className="form-group mb-4" data-pg-collapsed>
              <label className="font-weight-bold" htmlFor="tags">Tags (optional)</label>
              <input type="text" className="form-control" id="tags" name="tags" data-role="tagsinput" />
            </div>
            <div className="form-group mb-4">
              <div className="custom-file">
                <input type="file" className="custom-file-input" id="custom-file1" />
                <label className="custom-file-label" htmlFor="custom-file1">Upload cover image</label>
              </div>
            </div>
            <button type="submit" className="btn btn-brand btn-plain mt-3">Publish article</button>
          </form>
        </div>
      </div>
    </section>
  );
}

export default CreateArticle;
