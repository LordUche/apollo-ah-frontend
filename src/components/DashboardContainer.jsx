import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { getDashboard, getDashboardSummary } from '../selectors/dashboardSelector';
import { getProfile } from '../selectors/profileSelector';
import Dashboard from '../views/Dashboard';
import ProfileHeaderContainer from './ProfileHeaderContainer';
import { getDashboardArticles } from '../actions/dashboardAction';
import CommentLoader from '../views/CommentLoader';

/**
 * Container component for the Dashboard view
 * @export
 * @class DashboardContainer
 * @extends {Component}
 */
class DashboardContainer extends Component {
  async componentDidMount() {
    const { profile: { id } } = this.props;
    const { getDashboardArticles: fetchDashboardArticles } = this.props;
    await fetchDashboardArticles(id);
  }

  render() {
    const { dashboard, summary } = this.props;
    return (
      <>
        <ProfileHeaderContainer activePage="DASHBOARD" />
        {
          dashboard.loading
            ? <CommentLoader />
            : <Dashboard summary={summary} dashboard={dashboard} />
        }
      </>
    );
  }
}

DashboardContainer.propTypes = {
  dashboard: PropTypes.object.isRequired,
  summary: PropTypes.object.isRequired,
  profile: PropTypes.object.isRequired,
  getDashboardArticles: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector(
  {
    dashboard: getDashboard,
    summary: getDashboardSummary,
    profile: getProfile,
  }
);
const mapDispatchToProps = dispatch => ({
  getDashboardArticles: payload => dispatch(getDashboardArticles(payload)),
});
export default connect(mapStateToProps, mapDispatchToProps)(DashboardContainer);
