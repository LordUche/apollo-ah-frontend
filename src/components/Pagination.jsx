/* eslint-disable no-shadow */
import React, { Component } from 'react';
import ReactPaginate from 'react-paginate';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import PropTypes from 'prop-types';
import { getBookmarkedArticles } from '../actions/getBookmark';
import * as selectors from '../selectors/getBookmarkSelector';

/**
 * @description - paginates articles
 * @param {props} getArticles - action to get all articles
 * @param {props} last - represets the last page after articles are splitted by pages
 * @return {JSX}
 */
class Pagination extends Component {
  /**
   * @description - handles each click in the pagination bar
   * @param {props} selected - represents the page we are selecting
   * @return {void}
   */
  handlePageClick = async ({ selected }) => {
    const { getBookmarkedArticles } = this.props;
    const articlesSize = 6;
    // adding 1 because selected start from 0 while our articles page start from 1
    await getBookmarkedArticles(selected + 1, articlesSize);
    // Scroll the window to the top/beginning of articles for better user experience
    window.scrollTo({
      top: 10,
      behavior: 'smooth',
    });
  }

  render() {
    const { lastPage } = this.props;
    return (
      <div className="container">
        <ReactPaginate
          previousLabel={<i className="fas fa-chevron-left" />}
          nextLabel={<i className="fas fa-chevron-right" />}
          breakLabel="..."
          breakClassName="pagination-list"
          breakLinkClassName="pagination-anchor-tag"
          pageCount={lastPage}
          marginPagesDisplayed={2}
          pageRangeDisplayed={5}
          containerClassName="pagination justify-content-center mt-5"
          pageClassName="pagination-list"
          pageLinkClassName="pagination-anchor-tag"
          onPageChange={this.handlePageClick}
          subContainerClassName="pages pagination"
          activeClassName="active"
        />
      </div>
    );
  }
}

Pagination.propTypes = {
  lastPage: PropTypes.number.isRequired,
  getBookmarkedArticles: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector(
  {
    lastPage: selectors.getLastPage,
  }
);

export default connect(mapStateToProps, { getBookmarkedArticles })(Pagination);
