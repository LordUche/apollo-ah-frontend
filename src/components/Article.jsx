import React from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import ArticleBody from '../views/ArticleBody';
import CommentsContainer from './CommentsContainer';

/**
 *
 *
 * @class Article
 * @extends {React.Component}
 */
class Article extends React.Component {
  /**
   *
   * @static - Validating proptypes
   * @memberof Article
   */
  static propTypes = {
    match: PropTypes.object.isRequired,
  }

  /**
   * Creates an instance of Article.
   * @param {*} props - Props passed down from the BrowserRouter
   * @memberof Article
   */
  constructor(props) {
    super(props);
    this.state = {
      article: {},
      bookmarked: false,
    };
  }

  /**
   *
   * @todo - Move the fetch article functionality to redux store
   * @memberof Article
   */
  componentDidMount = async () => {
    const { match } = this.props;
    const { slug } = match.params;
    const result = await axios.get(`${process.env.API_BASE_URL}/articles/${slug}`);
    this.setState({ article: result.data.data });
  }

  /**
   * @todo - employ javascript debounce to avoid calling API on repeated clicks
   * @method bookmarkArticle - The bookmark article functionality
   * @memberof Article
   */
  bookmarkArticle = () => {
    const { bookmarked } = this.state;
    this.setState({ bookmarked: !bookmarked });
  };

  /**
   *
   *
   * @returns {JSX} The JSX representation of the Article component
   * @memberof Article
   */
  render() {
    const { article, bookmarked } = this.state;
    return (
      <div>
        <main className="main-body">
          <section className="min-vh-100">
            {article.image && (
              <div
                className="pg-empty-placeholder single-hero"
                style={{ backgroundImage: `url(${article.image})` }}
              />
            )}
            <div className="container">
              <div className="single-container">
                <ArticleBody
                  article={article}
                  bookmarkArticle={this.bookmarkArticle}
                  bookmarked={bookmarked}
                />
                <CommentsContainer />
              </div>
            </div>
            <div className="single-suggested-grp">
              <div className="container">
                <h4>Recommended for you</h4>
                <div className="row pg-empty-placeholder" />
              </div>
            </div>
          </section>
        </main>
      </div>
    );
  }
}

export default Article;
