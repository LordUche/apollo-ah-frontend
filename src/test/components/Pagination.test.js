import React from 'react';
import 'jest';
import { shallow } from 'enzyme';
// import reduxStore from '../../store';
import { Pagination } from '../../components/Pagination';

// const { store } = reduxStore;
const props = {
  getArticles: jest.fn(),
  last: 4,
  selected: 1
};
window.scrollTo = jest.fn();

describe('<Pagination Test Suite>', () => {
  it('It should render unconnected pagination succesfully', async (done) => {
    const wrapper = await shallow(<Pagination {...props} />);
    expect(wrapper).toMatchSnapshot();
    wrapper.instance().handlePageClick({ selected: 1 });
    expect(props.getArticles).toHaveBeenCalled();
    done();
  });
});
