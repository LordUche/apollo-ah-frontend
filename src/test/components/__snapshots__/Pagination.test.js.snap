// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination Test Suite> <Pagination> It should render connected pagination succesfully 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <MemoryRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "ra7mav",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "ra7mav",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Pagination)
        getArticles={[MockFunction]}
        last={4}
      >
        <Pagination
          description=""
          getArticles={[Function]}
          last={1}
        >
          <div
            className="container"
          >
            <PaginationBoxView
              activeClassName="active"
              breakClassName="pagination-list"
              breakLabel="..."
              breakLinkClassName="pagination-anchor-tag"
              containerClassName="pagination justify-content-center mt-5"
              disableInitialCallback={false}
              disabledClassName="disabled"
              marginPagesDisplayed={2}
              nextClassName="next"
              nextLabel={
                <i
                  className="fas fa-chevron-right"
                />
              }
              onPageChange={[Function]}
              pageClassName="pagination-list"
              pageCount={1}
              pageLinkClassName="pagination-anchor-tag"
              pageRangeDisplayed={5}
              previousClassName="previous"
              previousLabel={
                <i
                  className="fas fa-chevron-left"
                />
              }
              subContainerClassName="pages pagination"
            >
              <ul
                className="pagination justify-content-center mt-5"
              >
                <li
                  className="previous disabled"
                >
                  <a
                    aria-disabled="true"
                    onClick={[Function]}
                    onKeyPress={[Function]}
                    role="button"
                    tabIndex="0"
                  >
                    <i
                      className="fas fa-chevron-left"
                    />
                  </a>
                </li>
                <PageView
                  activeClassName="active"
                  key="0"
                  onClick={[Function]}
                  page={1}
                  pageClassName="pagination-list"
                  pageLinkClassName="pagination-anchor-tag"
                  selected={true}
                >
                  <li
                    className="pagination-list active"
                  >
                    <a
                      aria-current="page"
                      aria-label="Page 1 is your current page"
                      className="pagination-anchor-tag"
                      onClick={[Function]}
                      onKeyPress={[Function]}
                      role="button"
                      tabIndex="0"
                    >
                      1
                    </a>
                  </li>
                </PageView>
                <li
                  className="next disabled"
                >
                  <a
                    aria-disabled="true"
                    onClick={[Function]}
                    onKeyPress={[Function]}
                    role="button"
                    tabIndex="0"
                  >
                    <i
                      className="fas fa-chevron-right"
                    />
                  </a>
                </li>
              </ul>
            </PaginationBoxView>
          </div>
        </Pagination>
      </Connect(Pagination)>
    </Router>
  </MemoryRouter>
</Provider>
`;
